rules/store.go

```diff
@@ -0,0 +1,108 @@
+package rules
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+type AlertStore interface {
+	// SetAlerts stores the provided list of alerts for a rule
+	SetAlerts(key uint64, alerts []*Alert)
+	// GetAlerts returns a list of alerts for each rule in group,
+	// rule is identified by a hash of its config.
+	GetAlerts(key uint64) map[uint64]*Alert
+}
+
+type FileStore struct {
+	logger       log.Logger
+	alertsByRule map[uint64][]*Alert
+	//protects the `alertsByRule` map
+	stateMtx sync.RWMutex
+	path     string
+}
+
+func NewFileStore(l log.Logger, storagePath string) *FileStore {
+	s := &FileStore{
+		logger:       l,
+		alertsByRule: make(map[uint64][]*Alert),
+		path:         storagePath,
+	}
+	s.initState()
+	fmt.Println("starting alert store")
+
+	return s
+}
+
+// initState reads the state from file storage into the stateByRule map
+func (s *FileStore) initState() {
+	file, err := os.Open(s.path)
+	if err != nil {
+		level.Error(s.logger).Log("Error reading alerts state from file", err)
+	}
+	defer file.Close()
+
+	var stateByRule map[uint64][]*Alert
+	err = json.NewDecoder(file).Decode(&stateByRule)
+	if err != nil {
+		level.Error(s.logger).Log("Error reading alerts state from file", err)
+	}
+	if stateByRule == nil {
+		stateByRule = make(map[uint64][]*Alert)
+	}
+	s.alertsByRule = stateByRule
+}
+
+// GetAlerts returns the stored alerts for an alerting rule
+// Alert state is read from the in memory map which is populated during initialization
+func (s *FileStore) GetAlerts(key uint64) map[uint64]*Alert {
+	fmt.Println("reading alert state", "key", key)
+
+	s.stateMtx.RLock()
+	defer s.stateMtx.RUnlock()
+
+	restoredAlerts, ok := s.alertsByRule[key]
+	if !ok {
+		return nil
+	}
+	alerts := make(map[uint64]*Alert)
+	for _, alert := range restoredAlerts {
+		if alert == nil {
+			continue
+		}
+		h := alert.Labels.Hash()
+		alerts[h] = alert
+	}
+
+	return alerts
+}
+
+// SetAlerts updates the stateByRule map and writes state to file storage
+func (s *FileStore) SetAlerts(key uint64, alerts []*Alert) {
```

Comment:
should this method return errors and let the application decide what to do with it instead logging from inside of this module?

Reply:
I thought about this but could come not come up with what we can do if storing fails apart from logging the error. Rule evaluation should not fail even if storing the state fails. 

Is it better to return errors from this method and log it in the caller?

Reply:
I think it is fine to not fail the application, but you should let the application decide what to do with the error instead of swallowing it. This opens the door for the application to change their behaviour in the future without changing any api.